### PR TITLE
LSP Followup

### PR DESCRIPTION
**Pull request by norcalli**
Addressing things brought up in https://github.com/neovim/neovim/issues/11389 and beyond.

TODO:
- [x] reduce markdown previews of code.
- [x] refactor method triggers to `lsp.buf.*` methods
- [x] switch to v:lua and get rid of vim interface.
- [x] get rid of filetype config in favor of something from https://github.com/neovim/nvim-lsp
- [x] switch error messages to something which doesn't require `ENTER` because if an LSP goes crazy, it'll block neovim.
- [x] finish textDocument/rename
- [x] rename `builtin_callbacks` to `default_callbacks`
- [x] resolve callback at time of calling using default_callbacks instead of at client creation
- [x] Make hover/signatureHelp preview focusable so you can mess with it.
- [x] textDocument/references
- [x] Finish textDocument/rangeFormatting and textDocument/format
- [x] vim.lsp.util.set_qflist and vim.lsp.util.set_loclist
- [x] test more complex text edits because deletion is probably buggy.
- [x] test more complex text edits because deletion is probably buggy.
- [ ] Improve `update_tagstack`/location stuff.  from bfredl `update_tagstack seems like something that should be primitive. Maybe using 'tagfunc' option sidesteps the entire tagstack issue?`
- [ ] adjust offset encoding of parameters sent to callbacks.
- [ ] Fix completion
- [ ] take improvements from nvim-lsp
- [ ] check race condition in for_each_buffer_client. Should probably just skip closed clients.
About specific callbacks:
- [ ] For servers that only support full sync: we should throttle full buffer updates in some situations like global change.
- [ ] Simplify capability check in client.request (use a table or something)?
- [ ] Diagnostic code guessing guifg from guibg is oddly specific. Diagnostic code should just use LspDiagnostics{severity} which should be defined :hi default and thus overridable by user by standard means. A mechanism for generating bright/dark colors should be separate from that, and invoked together by user if they want to.

Feature requests:

- [ ] Support LocationLink as well as Location for definition etc. if server supports it (allows defining a peekDefinition mapping that previews the target location in a float)
- [ ] signature should highlight current param (I also would like it aligned to function start in code like vim-jedi does, but the info is not directly available from server)
- [ ] Support multiple Location[Link]s (by putting them in the loclist?)
- [ ] Allow overriding the symbolKind and completionItemKind mappings in server configs (if you prefer procedure to function, say; it's particularly useful for "less typical" servers like texlab)
- [ ] Provide a utility function to check if configured servers are running that can be used in the status line/airline/lightline to indicate if server is up or crashed.
- [ ] Support filterText for completion (to only show items matching the already typed text, either as prefix or as substring)

Related issues:
- [ ] improve interface for special types in Lua like an empty dict, because may cause problems for people (and me).
- [ ] add bitop library to core
- [ ] add native lua json functions?

### CODE REVIEW

**Comment by justinmk** [comment]
> switch error messages to something which doesn't require ENTER because if an LSP goes crazy, it'll block neovim.

Maybe just `:redraw` if the error message is longer than some reasonable length. That will skip hit-enter prompt.

**Comment by escorponox** [comment]
> switch error messages to something which doesn't require ENTER because if an LSP goes crazy, it'll block neovim.

You could use virtual text or floating windows like coc.nvim

image

**Comment by bfredl** [comment]
@escorponox We do this already. I think @norcalli meant errors like if the server returned an error from a request, or a malformatted return value or something.

File: runtime/lua/vim/lsp.lua [hunk 1]
```diff
@@ -26,6 +27,11 @@ local lsp = {
 -- TODO consider whether 'eol' or 'fixeol' should change the nvim_buf_get_lines that send.
 -- TODO improve handling of scratch buffers with LSP attached.
 
+local function err_message(...)
+  nvim_err_writeln(table.concat(vim.tbl_flatten{...}))
```
Comment (bfredl) [inline runtime/lua/vim/lsp.lua]: TODO (for myself): future `nvim_message` function should have "truncate" flag. I e show only the first 'cmdheight' line[s] in message area + some indicator, but store the full message in history.

**Comment by norcalli** [comment]
I was a bit overzealous in calling rename and formatting "finished." They work in many cases, but some edits are getting corrupted. I'm going to take a look at how other clients do it later.

**Comment by Seirdy** [comment]
Some documentation and example configs would also be appreciated. If said documentation exists elsewhere, links to those locations should suffice.

**Comment by norcalli** [comment]
This is how my new settings look for anyone curious to try it out. we just use the buf for most things:

```lua
local function language_client_setup(_)
  local mappings = {
    ["nK"]    = map_cmd [[lua vim.lsp.buf.hover()]];
    ["ngd"]   = map_cmd [[lua vim.lsp.buf.definition()]];
    ["ngr"]   = map_cmd [[lua vim.lsp.buf.rename()]];
    ["ngD"]   = map_cmd [[lua vim.lsp.util.show_line_diagnostics()]];
    ["n af"]  = map_cmd [[lua vim.lsp.buf.formatting()]];
    ["x af"]  = map_xmd [[lua vim.lsp.buf.range_formatting()]];
  }
  nvim.bo.omnifunc = "v:lua.vim.lsp.omnifunc"
  nvim_apply_mappings(mappings, { buffer = true; silent = true; })
end
```

I add it on `on_attach`. Formatting is close to being fixed.

**Comment by norcalli** [comment]
@bfredl @justinmk @h-michael alright, for a first pass, it's in a good state right now to be reviewed and merged. There are still things todo, but this fixes a lot of issues that people who are testing this are experiencing and adds some solid features.

So should we pause here and make another followup or should I continue working on it?

File: runtime/lua/vim/lsp/util.lua [hunk 2]
```diff
@@ -26,11 +26,85 @@ local function remove_prefix(prefix, word)
...
+    error("Invalid range: "..vim.inspect{A = A; B = B; #lines, new_lines})
   end
-  return bufnr
+  local prefix = ""
+  local suffix = lines[i_n]:sub(B[2]+1)
+  if A[2] > 0 then
+    prefix = lines[i_0]:sub(1, A[2])
+  end
+  new_lines = list_extend({}, new_lines)
+  if #suffix > 0 then
```
Comment (norcalli) [inline runtime/lua/vim/lsp/util.lua]: ordering here is very deliberate, btw

**Comment by norcalli** [comment]
And to prove that formatting works, I reformatted `eval.c`. it only took like 3 seconds even with all the extra copying I do in `set_lines` and compiled fine :D

**Comment by suessflorian** [comment]
Doing amazing work @norcalli 🙏

**Comment by h-michael** [comment]
@norcalli 
Does this have anything to do with this PR, but is there any clear reason to make the hover window appear above the cursor?

**Comment by norcalli** [comment]
@h-michael I'm pretty sure that code is mostly the same from your PR. So that seems like a better question for yourself? Do you find the position disagreeable?

**Comment by h-michael** [comment]
I intended to implement it to display under the cursor unless it protrudes. Okay, let's stop this topic here.

File: runtime/doc/lsp.txt [hunk 3]
```diff
@@ -508,43 +422,43 @@ vim.lsp.log_levels
...
-        " This is optional, but you may find it useful
-        autocmd CompleteDone * pclose
+  " Configure for python
+  autocmd Filetype python setl omnifunc=v:lua.vim.lsp.omnifunc
+  
+  " Or with on_attach
+  start_client {
+    ...
+    on_attach = function(client, bufnr)
+      vim.api.nvim_buf_set_option(bufnr, 'omnifunc', 'v:lua.vim.lsp.omnifunc')
```
Comment (norcalli) [inline runtime/doc/lsp.txt]: TODO me
```suggestion
      vim.bo(bufnr).omnifunc = 'v:lua.vim.lsp.omnifunc'
```

**Pull request merged**